### ISSUE TITLE
Add entropy regularization (e.g., `initial_entropy_coef`) to GRPOTrainer for improved exploration

### ISSUE THREAD

**Issue by lailanelkoussy** [huggingface/trl#5584]
### Feature request

I propose adding **explicit entropy regularization** to the GRPO training objective in TRL, with support for an **initial entropy coefficient and optional scheduling (decay over training steps)**.

Concretely, this would extend the current GRPO loss with an entropy term:

[
\mathcal{L} = \mathcal{L}_{GRPO} + \beta \cdot \mathcal{H}(\pi)
]

and expose user-facing parameters such as:

```python
GRPOConfig(
    initial_entropy_coef=0.02,
    final_entropy_coef=0.0,
    entropy_decay_steps=1000,
)
```

or a simpler version:

```python
GRPOConfig(
    entropy_coef=0.01,
)
```

This would provide a direct mechanism to control exploration at the **policy level**, rather than relying solely on sampling parameters (temperature, top-p).

**Related work / context:**

* Entropy regularization is standard in policy optimization methods such as PPO (e.g., Schulman et al., 2017)
* GRPO paper: https://arxiv.org/abs/2402.03300
* TRL implementation: https://github.com/huggingface/trl

### Motivation

Currently, exploration in GRPOTrainer is controlled indirectly via sampling parameters (e.g., temperature, top-p) and KL regularization.

In practice, this can lead to **very low exploration from the first training steps**, especially when:

* the base model has sharp logits
* KL penalty is non-negligible
* group samples become nearly identical

Observed issues include:

* very low diversity within GRPO groups
* near-zero reward variance across candidates
* weak or noisy learning signal
* early entropy collapse

Adjusting sampling parameters helps, but:

* it is indirect and unstable
* it does not affect the optimization objective itself
* it can still fail when KL constraints dominate

KL regularization toward a reference model does not reliably preserve entropy:

* if the reference policy is low-entropy → exploration collapses
* if KL is strong → deviations (including exploratory ones) are penalized

Adding entropy regularization would:

* encourage exploration directly in the objective
* improve diversity within groups → better relative ranking signal
* reduce sensitivity to sampling hyperparameters
* stabilize early training dynamics

This is a standard and well-understood technique in RL, and its absence in GRPOTrainer makes exploration fragile in practice.

### Your contribution

I’d be happy to contribute a PR implementing this feature.

The change appears relatively localized (loss computation + config), and I can:

* add entropy term to the GRPO loss
* implement a simple linear decay schedule
* expose config parameters in `GRPOConfig`
* include basic logging (entropy over time)

**Comment by qgallouedec** [comment]
Thanks, any paper showing this setting improves the training?

**Comment by lailanelkoussy** [comment]
Thanks for the question — while I’m not aware of a GRPO-specific paper evaluating entropy regularization, there is strong evidence from the broader policy gradient literature that supports this addition.

* Entropy regularization is widely used to prevent policies from becoming overly deterministic early and to improve exploration and convergence (e.g., PPO: https://arxiv.org/abs/1707.06347)
* It has been shown to improve exploration and stability in policy optimization methods (see overview: https://spinningup.openai.com/en/latest/algorithms/ppo.html)
* More recent work (e.g., *Entropy-Preserving Reinforcement Learning*, ICLR 2026) shows that policy gradient methods naturally suffer from entropy collapse during training, and that explicitly controlling entropy leads to improved performance and sustained exploration (https://machinelearning.apple.com/research/entropy-preserving-reinforcement-learning)
* Additionally, adaptive entropy scheduling has been proposed to better balance exploration vs exploitation compared to fixed coefficients (e.g., https://arxiv.org/abs/2003.01709)

Since GRPO is a policy gradient method, these findings should apply directly.

In my own experiments, I’m observing behavior consistent with this: **the very first rollout starts with near-zero reward (≈0), and training takes a long time to meaningfully improve**. Looking into it, this corresponds to **very low diversity in the sampled outputs from the start**, which suggests early entropy collapse and weak exploration.

This matches the failure mode described in the literature and is what motivates the proposal for adding entropy regularization (ideally with a decaying schedule) to stabilize early training.

**Comment by qgallouedec** [comment]
Well, those are good signs. I don't think that's enough yet to justify adding it to TRL, but as soon as a paper exists that uses this method, we should consider adding it

**Comment by albertvillanova** [comment]
Hey @lailanelkoussy, thanks for the detailed writeup! 

This is essentially the same request as #3320 ("Add Adaptive Entropy Control to GRPOTrainer"), which already has more discussion — including the empirical signal @qgallouedec asked about here (the 1-shot RLVR paper's ablation showing entropy loss lifts MATH500 from 70.8 → 74.8) and the note that verl ships entropy loss as standard. There's also an in-progress PR there, #3628, implementing static-then-adaptive entropy control. 

I'll close this as a duplicate so everything stays in one place — please follow / +1 on #3320, and your offer to contribute would be very welcome on #3628. (Side note: GRPO already supports entropy-based token masking via `top_entropy_quantile`, which is related but distinct from the entropy bonus you're proposing.)
